### ISSUE TITLE
Usage with redux

### ISSUE THREAD

**Issue by Ldoppea** [lavrton/react-konva#182]
Hi,

Sorry this is not really an issue, but a question.

A few weeks ago I started a project with Vue + Vuex + Vue-Konva and encountered some problems with layer management. And now I'm wondering if going React + Redux + React-Konva may fit my needs.

Here is the description of my problem:
- Take the star demo : https://codepen.io/lavrton/pen/myBPGo
- In my project I want to have similar behavior using React-Konva for displaying stars, and using Redux to store stars positions.
- Each time the dragmove ticks for a star, then the store is updated with new star position. This produce a lot of Redux mutations, but I wanted to have some other components on different layers to follow the star. (a bit like grouped elements, but without groups and on multiple layers and independent of each other)
- Like for the star demo, I want my dragged element to be temporary moved on a top layer. With Vuex, if I do that the konva way ( `shape.moveTo(layer)` ) then on next store mutation, the element will be drawn back to its origin layer and sometimes this results to shape displayed twice (one on each layer).

**Is this behavior the same with React-Konva**?

When I'm reading the documentation of each projects and engines (vue vs react), i'm pretty sure this will be the same result, but maybe I missed an option?

I'll try to create a sandbox project to try this next weekend, but just in case I ask here first.
Also if you want a demo for reproducing the problem I can make one (but not sure before next weekend).

Thanks

PS : maybe this is a duplicate of https://github.com/lavrton/react-konva/issues/179 but I am not sure.

**Comment by lavrton** [comment]
This case will be a bit tricky. You just need to store "isDragging" (or anything similar) property into your react store, and then in `render` function detect required layer for a shape by this property.

Demo: https://codesandbox.io/s/xv8kmkjkwp

**Comment by Ldoppea** [comment]
Thanks for your answer and the useful demo.

First, it's good news for me that React can do what I want 👍 

Then I tried to reproduce this sample with Vue-Konva as I did not try this solution. But seems like there are a few differences between both implementations and it does not work due to the way Vue-Konva orders mount and unmount/destroy events.

I won't ask you for support on Vue-Konva but maybe you can explain me why it is working in React? 😃 
I investigated your sample and events happen in that order : 
- drag start
- the store is updated (and so components)
- component on base layer calls `componentWillUnmount`
- component on drag layer calls `componentDidMount` (and so startDrag)
- `dragend` is called on base layer component due to `this._stage.destroy();` on https://github.com/lavrton/react-konva/blob/master/src/index.js

My understanding of React lifecycle is that `componentWillUnmount` should call `this._stage.destroy();` directly. But it appears this is not the case ( `componentDidMount` is called between them ) and this is why your code is working (otherwise `dragend` would be called first and so `isDragging` would be set to `false` before the `componentDidMount` resulting with no `dragStart`).

Why isn't `this._stage.destroy();` called directly after `componentWillUnmount`? Did I miss some React mechanisms?

**Comment by lavrton** [comment]
`this._stage.destroy();` is not calling here. This function is for stage destroy but here we are not destroying a stage.


The order is:
1. Drag start
2. The target is updated with `isDragging: true`
3. Since the store is changed, React is doing re-render
4. On render function `<Target>` is moved into another layer
5. That action will trigger two effects
6. First effect - Target is removed from original layer so `componentWillUnmount` for `<Target>` is called. Also dragging of konva node will be stopped (since its removing)
7. Second effect - new component `<Target >` is added to drag layer. But we need to call `startDrag` manually, because it is a new node, so it is not dragging.

I updated demo a bit: https://codesandbox.io/s/xv8kmkjkwp

**Comment by Ldoppea** [comment]
Thanks again for your answer and awesome demo.

With this demo I found a similar solution working for Vue, but it requires a small modification to be made on vue-konva (a `beforeDestroy` event that should be replaced by `destroyed`). I'll discuss with its creator about potential side-effects and see if the modification can be made. If so I'll port your demo using vue and post it there as a thank you.

Also, not really a bug but this might interest you : I noticed that when adding a shadow with opacity to the draggable component, there will be some tilting in the shadow at the end of the drag&drop.  Here is a fork reproducing this behavior : https://codesandbox.io/s/zxny6w50o4
As I don't really need shadows in my project, this is not an issue for me.

Thanks again
